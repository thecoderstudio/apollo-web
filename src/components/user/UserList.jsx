import React from 'react';
import axios from 'axios';
import UserListItem from './UserListItem';


export default class UserList extends React.PureComponent {
  constructor(props) {
    super(props);
    this.fetchUsers = this.fetchUsers.bind(this);
    this.state = {
      users: []
    }
  }

  componentDidMount() {
    this.fetchUsers();
  }

  fetchUsers() {
    axios.get(
      `${process.env.APOLLO_HTTP_URL}user`,
      { withCredentials: true }
    ).then(res => {
      if (res.status === 200) {
        this.setState({
          users: res.data
        });
      }
    });
  }

  render() {
    return (
      <div>
        {this.state.users.map(user => (
          <UserListItem key={user.id} user={user} />
        ))}
      </div>
    );
  }
}
