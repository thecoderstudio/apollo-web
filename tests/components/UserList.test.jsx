import React from 'react';
import renderer from 'react-test-renderer';
import axios from 'axios';
import waitForExpect from 'wait-for-expect';
import UserList from '../../src/components/user/UserList';

jest.mock('axios');

function getComponent() {
  return renderer.create(<UserList />);
}

describe("user list", () => {
  beforeEach(() => {
    process.env = {
      APOLLO_HTTP_URL: 'http://localhost:1234/'
    };
  });

  it("renders correctly", async () => {
    const users = [
      {
        username: 'test',
        role: null
      },
      {
        username: 'admin',
        role: {
          name: 'admin'
        }
      }
    ];
    axios.get.mockResolvedValue({
      status: 200,
      data: users
    });

    const component = getComponent()
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    await waitForExpect(() => {
      expect(axios.get).toHaveBeenCalled();
    });
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("doesn't try to render users on error", async () => {
    axios.get.mockResolvedValue({
      status: 500,
      data: { detail: "generic error" }
    });

    const component = getComponent()
    let tree = component.toJSON();

    await waitForExpect(() => {
      expect(axios.get).toHaveBeenCalled();
    });
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

  });
});
