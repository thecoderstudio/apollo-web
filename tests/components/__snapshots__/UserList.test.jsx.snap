// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user list doesn't try to render users on error 1`] = `
<div>
  <h2
    className="sc-fznKkj hRHxK"
  >
    Users
  </h2>
  <div
    className="sc-fznZeY SPqpd"
  />
</div>
`;

exports[`user list renders correctly 1`] = `
<div>
  <h2
    className="sc-fznKkj hRHxK"
  >
    Users
  </h2>
  <div
    className="sc-fznZeY SPqpd"
  />
</div>
`;

exports[`user list renders correctly 2`] = `
<div>
  <h2
    className="sc-fznKkj hRHxK"
  >
    Users
  </h2>
  <div
    className="sc-fznZeY SPqpd"
  >
    <div
      className="sc-fzoLsD ffqgCm"
    >
      <p
        className="sc-fzpans kEnlFJ"
      >
        test
      </p>
      <button
        className="sc-AxjAm sc-AxirZ sc-fznyAO iiOVWb"
        onClick={[Function]}
      >
        delete
      </button>
    </div>
    <div
      className="sc-fzoLsD ffqgCm"
    >
      <p
        className="sc-fzpans kEnlFJ"
      >
        admin
      </p>
      <p
        className="sc-fzplWN dkHTim"
      >
        admin
      </p>
    </div>
  </div>
</div>
`;
